### PR TITLE
DROOLS-4208: [DMN Designer] Documentation - Widget for links persisting links

### PR DESCRIPTION
**Pull request by danielzhe**
DROOLS-4209: [DMN Designer] Documentation - Persist links into extension elements

This PR is part of an ensemble.
https://github.com/kiegroup/drools/pull/2431
https://github.com/kiegroup/kie-wb-common/pull/2739

documentations links demo

### CODE REVIEW

File: kie-dmn/kie-dmn-backend/src/main/java/org/kie/dmn/backend/marshalling/v1_2/xstream/DMNExternalLinkConverter.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    @Override
+    protected DMNModelInstrumentedBase createModelObject() {
+        return new TDMNExternalLink();
+    }
+
+    @Override
+    public boolean canConvert(final Class aClass) {
+        return aClass.equals(DMNExternalLink.class) || aClass.equals(TDMNExternalLink.class);
```
Comment (jomarko) [inline kie-dmn/kie-dmn-backend/src/main/java/org/kie/dmn/backend/marshalling/v1_2/xstream/DMNExternalLinkConverter.java]: Could we please use `Objects.equals` for comparing? It doesn't throw an exception if `aClass` is null.

File: kie-dmn/kie-dmn-backend/src/main/java/org/kie/dmn/backend/marshalling/v1_2/xstream/ExtensionElementsConverter.java [hunk 2]
```diff
@@ -69,40 +72,57 @@ public Object unmarshal(HierarchicalStreamReader reader, UnmarshallingContext co
             while (reader.hasMoreChildren()) {
                 reader.moveDown();
                 String nodeName = reader.getNodeName();
-                try {
-                    Object object = readItem(reader, context, null);
-                    if (object instanceof DMNModelInstrumentedBase) {
-                        ((KieDMNModelInstrumentedBase) object).setParent(obj);
-                        ((KieDMNModelInstrumentedBase) obj).addChildren((KieDMNModelInstrumentedBase) object);
+                if (EXTERNAL_LINKS.equals(nodeName)) {
```
Comment (jomarko) [inline kie-dmn/kie-dmn-backend/src/main/java/org/kie/dmn/backend/marshalling/v1_2/xstream/ExtensionElementsConverter.java]: Could we please use `Objects.equals`?

File: kie-dmn/kie-dmn-backend/src/main/java/org/kie/dmn/backend/marshalling/v1_2/xstream/ExtensionElementsConverter.java [hunk 3]
```diff
@@ -124,8 +144,11 @@ public boolean canConvert(Class clazz) {
 
     @Override
     public void assignChildElement(Object parent, String nodeName, Object child) {
-        ExtensionElements id = (ExtensionElements)parent;
-        id.getAny().add(child);
+        ExtensionElements id = (ExtensionElements) parent;
+        if (nodeName.equals(DMN_EXTERNAL_LINK)) {
```
Comment (jomarko) [inline kie-dmn/kie-dmn-backend/src/main/java/org/kie/dmn/backend/marshalling/v1_2/xstream/ExtensionElementsConverter.java]: Could we please use `Objects.equals`?

**Review by jomarko — COMMENTED** [review]
Just few comments about my favorite theme `Objects.equals`

**Review by manstis — COMMENTED** [review]
@danielzhe I took a look, but TBH, I'm not sure the generated XML is correct.

Our change adds this:
```
<dmn:extensionElements>
  <dmn:children></dmn:children>
  <dmn:additionalAttributes></dmn:additionalAttributes>
  <dmn:externalLinks>
    <dmn:DMNExternalLink linkDescription="cheese" url="http://cheese.jpg"></dmn:DMNExternalLink>
    <dmn:DMNExternalLink linkDescription="flan" url="http://flan.jpg"></dmn:DMNExternalLink
  </dmn:externalLinks>
</dmn:extensionElements>
```
Whereas I've seen this from other tooling:
```
<dmn:extensionElements>
  <xxx:attachment id="uuid1" name="cheese" url="http://cheese.jpg" ... />
  <xxx:attachment id="uuid2" name="flan" url="http://flan.jpg" ... />
</dmn:extensionElements>
```
I question:

1. Inclusion of `dmn:children`
2. Inclusion of `dmn:additionalAttributes`
3. Use of `dmn` namespace on `externalLinks` (`kie` would probably be better)
4. Use of `dmn` namespace on `DMNExternalLink` (`kie` would probably be better)

I've added @tarilabs on this PR review as I suspect he has more knowledge about how extension elements should operate - and more importantly persist.

**Comment by danielzhe** [comment]
@manstis I think the inclusion of `children` and `additionalAttributes` was not introduced in this PR. Let's wait for @tarilabs .

**Comment by tarilabs** [comment]
> Let's wait for @tarilabs

I'm on PTO as per calendar and auto reply and no PC access. Is this st @baldimir or @etirelli can provide feedback to in my absence pls?

**Comment by manstis** [comment]
@etirelli @baldimir I've added you to review this PR and consider my [comment](https://github.com/kiegroup/drools/pull/2431#pullrequestreview-259460339) as suggested by Matteo in his (PTO) absence. It's like only one of you may need to look so please feel free to remove whoever is not needed.

File: kie-dmn/kie-dmn-backend/src/main/java/org/kie/dmn/backend/marshalling/v1_2/xstream/ExtensionElementsConverter.java [hunk 4]
```diff
@@ -69,40 +73,57 @@ public Object unmarshal(HierarchicalStreamReader reader, UnmarshallingContext co
             while (reader.hasMoreChildren()) {
                 reader.moveDown();
                 String nodeName = reader.getNodeName();
-                try {
-                    Object object = readItem(reader, context, null);
-                    if (object instanceof DMNModelInstrumentedBase) {
-                        ((KieDMNModelInstrumentedBase) object).setParent(obj);
-                        ((KieDMNModelInstrumentedBase) obj).addChildren((KieDMNModelInstrumentedBase) object);
+                if (Objects.equals(EXTERNAL_LINKS, nodeName)) {
```
Comment (jomarko) [inline kie-dmn/kie-dmn-backend/src/main/java/org/kie/dmn/backend/marshalling/v1_2/xstream/ExtensionElementsConverter.java:76]: This new if then else is not covered much, do you think could we increase the coverage slightly?

https://sonarcloud.io/component_measures?id=org.drools%3Adrools&metric=new_coverage&pullRequest=2431&selected=org.drools%3Adrools%3Akie-dmn%2Fkie-dmn-backend%2Fsrc%2Fmain%2Fjava%2Forg%2Fkie%2Fdmn%2Fbackend%2Fmarshalling%2Fv1_2%2Fxstream%2FExtensionElementsConverter.java&view=list

**Review by jomarko — COMMENTED** [review]
I checked the coverage also, is good, just one block that has not trivial logic not covered, could you have a look @danielzhe ?

**Comment by danielzhe** [comment]
https://github.com/kiegroup/kie-wb-common/pull/2739#issuecomment-509937783
@manstis @jomarko So I think I have to change the `ExtensionElementsConverter` to skip `children` and `additionalAttributes`. Do you guys know any side effect of this?

**Comment by manstis** [comment]
@danielzhe I think you may need to reconsider your use of extension elements. As suggested on gChat please look at this [test](https://github.com/kiegroup/drools/blob/master/kie-dmn/kie-dmn-backend/src/test/java/org/kie/dmn/backend/marshalling/v1_1/UnmarshalMarshalTest.java#L123) to see how extensions are declared and used. Whilst this test is for v1_1 you can copy the test to the v1_2 marshaller test and with minor modifications see it work there too.

**Comment by danielzhe** [comment]
This PR is not necessary anymore, since all work is being done now in DMN side.

**Pull request closed without merging**